src/Psl/DateTime/Timestamp.php

```diff
@@ -0,0 +1,273 @@
+     * ```
+     *
+     * @param string $raw_string The date and time string to parse.
+     * @param null|DateStyle $date_style The style for the date portion of the string. If null, a default style is used.
+     * @param null|TimeStyle $time_style The style for the time portion of the string. If null, a default style is used.
+     * @param null|Timezone $timezone Optional timezone for parsing. If null, uses the system's default timezone.
+     * @param null|Locale $locale Optional locale for parsing. If null, uses the system's default locale.
+     *
+     * @throws Exception\RuntimeException If the parsing process fails.
+     *
+     * @return static Returns an instance of {@see Timestamp} representing the parsed date and time.
+     *
+     * @see TemporalInterface::toString()
+     *
+     * @psalm-mutation-free
+     */
+    public static function fromString(string $raw_string, null|DateStyle $date_style = null, null|TimeStyle $time_style = null, null|Timezone $timezone = null, null|Locale $locale = null): static
+    {
+        /** @psalm-suppress MissingThrowsDocblock */
+        return self::fromParts(
+            Internal\parse(raw_string: $raw_string, date_style: $date_style, time_style: $time_style, timezone: $timezone, locale: $locale)
+        );
+    }
+
+    /**
+     * Returns this Timestamp instance itself, as it already represents a timestamp.
+     *
+     * @psalm-mutation-free
+     */
+    public function getTimestamp(): self
+    {
+        return $this;
+    }
+
+    /**
+     * Returns the {@see Timestamp} parts (seconds, nanoseconds).
+     *
+     * @return array{int, int<0, 999999999>}
+     *
+     * @psalm-mutation-free
+     */
+    public function toParts(): array
+    {
+        return [$this->seconds, $this->nanoseconds];
+    }
+
+    /**
+     * Returns the number of seconds since the Unix epoch represented by this timestamp.
+     *
+     * @return int Seconds since the epoch. Can be negative for times before the epoch.
+     *
+     * @psalm-mutation-free
+     */
+    public function getSeconds(): int
+    {
+        return $this->seconds;
+    }
+
+    /**
+     * Returns the nanoseconds part of this timestamp.
+     *
+     * @return int<0, 999999999> The nanoseconds part, ranging from 0 to 999999999.
+     *
+     * @psalm-mutation-free
+     */
+    public function getNanoseconds(): int
+    {
+        return $this->nanoseconds;
+    }
+
+    /**
+     * Adds the specified duration to this timestamp object, returning a new instance with the added duration.
+     *
+     * @throws Exception\UnderflowException If adding the duration results in an arithmetic underflow.
+     * @throws Exception\OverflowException If adding the duration results in an arithmetic overflow.
+     *
+     * @psalm-mutation-free
+     */
+    public function plus(Duration $duration): static
```

Comment:
could you explain further? i don't see what issue might happen here? 🤔

Reply:
I think it comes down to this:

Currently, it is only possible to plus() or minus() a duration on Timestamp, but there is no way for adding years, months, ... since this is not part of Duration.

So currently that logic lives inside DateTime (which uses a Timestamp internally). On top of DateTime, there are added extra methods to add years and months specifically. At that location, the additional logic for calculating days in months / years are added. 

So I was thinking : If DateTime uses a Timestamp internally: why not make adding years / months part of timestamp instead so that is possible too. That way, the logic in DateTime could be a proxy to the methods in Timestamp.


Optionally, the Duration could be extended to years / months if the total seconds logic took a Timestamp as parameter to base calculations upon. I also see that's not part of brick-time, so I suppose that is not the best idea there. However it is a bit strange that it's not possible to `plus(Duration::years(1)` but it is possible to `plusYears(1)` from a user's perspective.

Reply:
we won't be able to do `Duration::years(1)`, the main difference between `Period` in java and `Duration` in psl is that `Duration` could be converted to a total number of seconds ( or milliseconds, nanoseconds .. etc ),. while the `Period` object in java can't, it is mearly a data object, so if you do +500 days on a two year period, it won't be 3 years and 135 days, but 2 years and 500 days, as it won't be able to convert the days to proper months / years without having a reference timestamp. e.g:

```java
import java.time.*;
import java.time.format.DateTimeFormatter;
import java.time.format.FormatStyle;
import java.time.temporal.ChronoUnit;
import java.time.temporal.TemporalAdjusters;


Period period = Period.ofYears(2).withDays(500);

System.out.println("👋 Period: " + period);
```

and the reason we want `Duration` to be convertable to seconds and fractions of seconds is that we want to use it for timeouts and such, `Period` in java cannot be used for this purpose, unless you have a reference point.

---

as for moving the logic of adding/removing years/months to `Timestamp`, that is also not possible, the reason being `Timestamp` could represent a monotomic time, adding / removing years from it makes no sense, as we do not have an actual timestamp reference to decide when a leap year should happen.

Reply:
Thanks for the explanation. Maybe a few remarks til I stop about this :) : 

* Brick's date-time package come shipped with `Duration` which is pretty much the same as what we got here. On top of that, it has a `Period` which can be used for date-ranges like years months. There the  difference is made between plusPeriod and plusDuration. it might make sense to prepare the classes for that as well?


*  About the monotomic timestamps : Does it make sense to have both monotomic and reguler Timestamps in the same class? I can imagine if you are doing performance testing, you don't really want to add days to the results there? Cause this would add days to unix epoch most likely?